designs/warmreplicas.md

```diff
@@ -0,0 +1,151 @@
+### Design
+1. Add a new interface `WarmupRunnable` that allows for leader election runnables to specify
+behavior when manager is not in leader mode. This interface should be as follows:
+```go
+type WarmupRunnable interface {
+    NeedWarmup() bool
+    GetWarmupRunnable() Runnable
+}
+```
+
+2. Controllers will implement this interface to specify behavior when the manager is not the leader.
+Add a new controller option `ShouldWarmupWithoutLeadership`. If set to true, then the main
+controller runnable will not start sources, and instead rely on the warmup runnable to start sources
+The option will be used as follows:
+```go
+type Controller struct {
+    // ...
+
+    // ShouldWarmupWithoutLeadership specifies whether the controller should start its sources
+    // when the manager is not the leader.
+    // Defaults to false, which means that the controller will wait for leader election to start
+    // before starting sources.
+    ShouldWarmupWithoutLeadership *bool
+
+    // ...
+}
+
+type runnableWrapper struct {
+    startFunc func (ctx context.Context) error
+}
+
+func(rw runnableWrapper) Start(ctx context.Context) error {
+    return rw.startFunc(ctx)
+}
+
+// NeedWarmup implements WarmupRunnable
+func (c *Controller[request]) NeedWarmup() bool {
+    if c.ShouldWarmupWithoutLeadership == nil {
+        return false
+    }
+    return c.ShouldWarmupWithoutLeadership
+}
+
+// GetWarmupRunnable implements WarmupRunnable
+func (c *Controller[request]) GetWarmupRunnable() Runnable {
+    return runnableWrapper{
+        startFunc: func (ctx context.Context) error {
+            if !c.ShouldWarmupWithoutLeadership {
+                return nil
+            }
+
+            // pseudocode
+            for _, watch := c.startWatches {
+                watch.Start()
+                // handle syncing sources
+            }
+        }
+    }
+}
+```
+
+3. Add a separate runnable category for warmup runnables to specify behavior when the
+manager is not the leader. [ref](https://github.com/kubernetes-sigs/controller-runtime/blob/v0.20.2/pkg/manager/runnable_group.go#L55-L76)
+```go
+type runnables struct {
+    // ...
+
+	LeaderElection *runnableGroup
+    Warmup *runnableGroup
+
+    // ...
+}
+
+func (r *runnables) Add(fn Runnable) error {
+	switch runnable := fn.(type) {
+    // ...
+    case WarmupRunnable:
+        if runnable.NeedWarmup() {
+            r.Warmup.Add(runnable.GetWarmupRunnable(), nil)
+        }
+
+        // fallthrough to ensure that a runnable that implements both LeaderElection and
+        // NonLeaderElection interfaces is added to both groups
+        fallthrough
+	case LeaderElectionRunnable:
+		if !runnable.NeedLeaderElection() {
+			return r.Others.Add(fn, nil)
+		}
+		return r.LeaderElection.Add(fn, nil)
+    // ...
+	}
+}
+```
+
+4. Start the non-leader runnables during manager startup.
+```go
+func (cm *controllerManager) Start(ctx context.Context) (err error) {
+    // ...
+
+    // Start the warmup runnables
```

Comment:
At which exact place in the current Start func would this go? (what do we do directly before/after)

Reply:
Initially I thought before we start the caches but that would mean that we can not respect the startup timeout in the controller so I guess it has to come after

Reply:
This would come [here](https://github.com/kubernetes-sigs/controller-runtime/blob/bd9c786e30178e26be5f5a5756dbc120c11f2e0e/pkg/manager/internal.go#L441) after non-leader election runnables and right before we start the leader elector runnables